ambry-cloud/src/main/java/com/github/ambry/cloud/RecoveryThread.java

```diff
@@ -0,0 +1,182 @@
+
+  /**
+   * Persists token to cloud in each replication cycle
+   * @param remoteReplicaInfo Remote replica info object
+   * @param exchangeMetadataResponse Metadata object exchanged between replicas
+   */
+  @Override
+  public void advanceToken(RemoteReplicaInfo remoteReplicaInfo, ExchangeMetadataResponse exchangeMetadataResponse) {
+    // Advance in-memory token
+    RecoveryToken oldToken = (RecoveryToken) remoteReplicaInfo.getToken();
+    super.advanceToken(remoteReplicaInfo, exchangeMetadataResponse);
+    // truncate previous token in-place and persist in-place
+    RecoveryToken token = (RecoveryToken) remoteReplicaInfo.getToken();
+    if (token == null) {
+      logger.error("Null token for replica {}", remoteReplicaInfo);
+      return;
+    }
+    if (token.equals(oldToken)) {
+      logger.trace("Not persisting recovery token as it is unchanged, oldToken = {}, newToken = {}", oldToken, token);
+      return;
+    }
+    if (token.isEndOfPartition()) {
+      logger.info("Completed recovering partition {}", remoteReplicaInfo.getReplicaId().getPartitionId().getId());
+      token.endRecovery();
+    }
+    logger.trace("replica = {}, token = {}", remoteReplicaInfo, token);
+    String tokenFilename = recoveryManager.getRecoveryTokenFilename(remoteReplicaInfo);
+    logger.trace("Writing recovery token to disk at {}", tokenFilename);
+    truncateAndWriteToFile(tokenFilename, token.toString());
+  }
+
+  /**
+   * Returns a cached file-desc or creates a new one
+   * @param filePath File system path
+   * @param options Options to use when opening or creating a file
+   * @return File descriptor
+   */
+  protected SeekableByteChannel getFd(String filePath, EnumSet<StandardOpenOption> options) {
+    FileDescriptor fileDescriptor = (FileDescriptor) fileDescriptorCache.getObject(filePath);
+    if (fileDescriptor == null) {
+      // Create parent folders
+      Path directories = Paths.get(filePath.substring(0, filePath.lastIndexOf(File.separator)));
+      try {
+        Files.createDirectories(directories);
+      } catch (IOException e) {
+        logger.error("Path = {}, Error creating folders = {}", directories, e.toString());
+        return null;
+      }
+      // Create file
+      try {
+        fileDescriptor = new FileDescriptor(Files.newByteChannel(Paths.get(filePath), options));
+      } catch (IOException e) {
+        logger.error("Path = {}, Options = {}, Error creating file = {}", filePath, options, e.toString());
+        return null;
+      }
+      // insert into cache
+      fileDescriptorCache.putObject(filePath, fileDescriptor);
+    }
+    return fileDescriptor.getSeekableByteChannel();
+  }
+
+  /**
+   * Truncates a file and then writes to it.
+   * Creates the file if absent.
+   * @param filePath Path of the file in the system
+   * @param text Text to append
+   * @return True if append was successful, false otherwise
+   */
+  protected boolean truncateAndWriteToFile(String filePath, String text) {
```

Comment:
nit: Return type could be `void` since it is not used?

Reply:
This is fine. A function must always return something, the caller can choose to use it or not.